feat: restart repo with PureEngineClient

tzdybal opened the pull request:
# Pure Engine API-based Execution Client for Rollkit

## Overview

This PR provides a complete reimplementation of the go-execution-evm package, with both code and documentation recreated from scratch. It implements the `execution.Executor` interface from the Rollkit `feature/exec_api` branch.

Resolves rollkit/go-execution-evm#7.
Resolves rollkit/go-execution-evm#9.
Fixes previously closed, but still broken: rollkit/go-execution-evm#3, rollkit/go-execution-evm#4, rollkit/go-execution-evm#5, rollkit/go-execution-evm#6.

## Key Changes

- **Complete Reimplementation**: Rebuilt the entire codebase and documentation to create a clean, efficient Engine API-based execution client
- **Rollkit Integration**: Designed specifically to work with the `feature/exec_api` branch in Rollkit
- **Comprehensive Testing**: Test suite verifies correct behavior by querying Reth using Eth API after each operation
- **Docker Integration**: Docker Compose setup is used directly in unit tests, ensuring both the compose configuration and genesis files are properly tested

## Technical Highlights

- Implements a `PureEngineClient` that uses both Engine API and standard Ethereum JSON-RPC API
- Maintains payload ID between calls for consistent block production
- Requires proper post-merge hardfork configurations in genesis
- Uses Docker-based setup with Reth for both development and testing

This implementation provides a clean, standards-compliant way to connect Rollkit to Ethereum execution clients.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **Documentation**
  - Updated product documentation with detailed sections on the new client implementation, genesis configuration, payload management, and API usage, including a refreshed deployment diagram.

- **New Features**
  - Enhanced the execution client for more robust transaction processing and state management.
  - Updated blockchain genesis settings to ensure a more consistent chain initialization.
  - Streamlined container setup with a new service image and simplified initialization commands.

- **Chores**
  - Removed legacy build targets and command-line interfaces to deliver a cleaner, more focused user experience.
  - Updated Go version and dependencies for improved compatibility and features.
  - Removed integration tests and mock implementations to refine the testing strategy.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

tzdybal committed:
test: run common test suite

tzdybal committed:
wip

tzdybal committed:
feat: update TxInjector

tzdybal committed:
build(deps): update go-execution dependency to latest version

Upgraded `github.com/rollkit/go-execution` from `fccb99951758` to `dc52ad420afc` in both `go.mod` and `go.sum`. This ensures compatibility with the latest changes and improvements in the dependency. No other dependencies were affected.

tzdybal committed:
feat: new Engine API approach

tzdybal committed:
feat: engine API that kind of works...

For some reason first payload cannot be used

tzdybal committed:
THIS WORKS! IT WAS GENESIS FILE ALL THE TIME!

Cleanup in next commits

tzdybal committed:
refactor: clean up test and engine code, improve block validation checks

- Remove debug print statements and unused imports
- Enhance block validation checks in test cases with more precise assertions
- Simplify block height and transaction tracking logic
- Remove commented-out code and unnecessary debug logging

tzdybal committed:
feat: enhance PureEngineClient with block retrieval and finalization

- Add Ethereum client to PureEngineClient for block information retrieval
- Implement `getBlockInfo` method to fetch block details by height
- Update `InitChain` and `SetFinal` methods to use dynamic block information
- Modify test cases to include `SetFinal` method calls
- Update command-line interface to pass Ethereum URL when creating client

tzdybal committed:
refactor: clean up test and engine code, improve block validation checks

- Remove debug print statements and unused imports
- Enhance block validation checks in test cases with more precise assertions
- Simplify block height and transaction tracking logic
- Remove commented-out code and unnecessary debug logging

tzdybal committed:
refactor: consolidate code in eexecution*.go files

- Merge `engine_test.go` and `execution_test.go` into a single test file
- Remove redundant test code and consolidate test logic
- Update function and method names for consistency
- Simplify transaction and block processing in test cases
- Remove unnecessary comments and debug logging

tzdybal committed:
refactor: restructure execution code and enhance testing framework

- Consolidate and refactor execution logic in `execution.go` and `execution_test.go`
- Remove deprecated integration tests and related files
- Update transaction handling to use byte arrays instead of custom types
- Improve error handling and logging in the execution client methods
- Adjust Go module dependencies and update to Go 1.24.0
- Remove unnecessary build commands from the Makefile
- Enhance test setup for better isolation and cleanup

tzdybal committed:
chore: fix linter errors

tzdybal committed:
refactor: update execution tests and docker-compose configuration

- Refactor `execution_test.go` to improve test setup and cleanup processes
- Change JWT filename from `testsecret.hex` to `jwt.hex` for clarity
- Update permissions for JWT directory and file to be more permissive
- Replace container setup with Docker Compose for better management
- Update Docker image version in `docker-compose.yml` from `v1.1.1` to `v1.2.1`
- Remove unnecessary volume bindings in Docker configuration

tzdybal committed:
docs: update README to reflect PureEngineClient implementation and usage

- Revise architecture section to introduce the `PureEngineClient` and its compatibility with the Engine API
- Detail how the Eth API is utilized for block information retrieval and transaction execution
- Explain the payload ID management and its significance in transaction processing
- Add new sections on development, testing, and reading genesis information with updated commands

tzdybal committed:
docs: enhance README with detailed genesis configuration for PureEngineClient

- Add a new section on genesis requirements for the `PureEngineClient`, outlining necessary hardfork settings
- Include an example of the required genesis configuration in JSON format
- Reorganize the section on how the Eth API is used, ensuring clarity on its role in transaction execution and block information retrieval

tzdybal commented on execution.go:
```diff
@@ -317,24 +250,18 @@ func (c *EngineAPIExecutionClient) setFinal(ctx context.Context, blockHash commo
-	if err != nil {
-		return err
-	}
+func (c *PureEngineClient) derivePrevRandao(blockHeight uint64) common.Hash {
+	return common.BigToHash(new(big.Int).SetUint64(blockHeight))
+}
```
Open a follow up issue for this comment.

tzdybal committed:
fix: update transaction timestamp handling in execution logic

- Modify `ExecuteTxs` method to accept a timestamp parameter instead of using the current time directly
- Update test cases in `execution_test.go` to use `time.Now()` for consistency in transaction execution timing

tzdybal commented on execution.go:
```diff
@@ -73,232 +73,165 @@
-			Timestamp:             uint64(timestamp.Unix()), //nolint:gosec // disable G115
-			Random:                c.derivePrevRandao(height),
+			Timestamp:             uint64(time.Now().Unix()), //nolint:gosec // disable G115
+			Random:                c.derivePrevRandao(blockHeight),
```
Fixed in 8f09ce6.

MSevey commented on execution.go:
```diff
@@ -73,232 +73,165 @@ func NewEngineAPIExecutionClient(
+// ExecuteTxs executes the given transactions at the specified block height and timestamp
+func (c *PureEngineClient) ExecuteTxs(ctx context.Context, txs [][]byte, blockHeight uint64, timestamp time.Time, prevStateRoot []byte) (updatedStateRoot []byte, maxBytes uint64, err error) {
+	var payloadResult engine.ExecutionPayloadEnvelope
+	// First tx is the serialized payload
+	firstTx := txs[0]
```
nit: worth validating the len of txs > 0 to avoid Panics due to developer errors.

tzdybal replied:
Follow-up in #35

MSevey reviewed (dismissed):
Nice 👏🏻

gupadhyaya commented on execution.go:
```diff
@@ -73,232 +73,165 @@ func NewEngineAPIExecutionClient(
 		engine.PayloadAttributes{
 			Timestamp:             uint64(genesisTime.Unix()), //nolint:gosec // disable G115
-			Random:                common.Hash{},
+			Random:                common.Hash{},              // TODO(tzdybal): this probably shouldn't be 0
```
some frameworks use blockhash as random (here it is genesis hash).

tzdybal replied:
Also part of https://github.com/rollkit/go-execution-evm/issues/34

gupadhyaya commented on execution.go:
```diff
@@ -73,232 +73,165 @@ func NewEngineAPIExecutionClient(
+			Random:                c.derivePrevRandao(blockHeight),
 			SuggestedFeeRecipient: c.feeRecipient,
-			Withdrawals:           []*types.Withdrawal{},
 			BeaconRoot:            &c.genesisHash,
```
should we use `prevStateRoot` instead of a constant genesisHash. This field is supposed to be used to store any vital consensus layer information.

gupadhyaya commented on execution.go:
```diff
@@ -73,232 +73,165 @@ func NewEngineAPIExecutionClient(
-			FinalizedBlockHash: prevBlockHash,
+			HeadBlockHash: blockHash,
+			SafeBlockHash: blockHash,
+			// FinalizedBlockHash: blockHash,
```
i guess this is on purpose to allow for SetFinal to set this?

tzdybal replied:
Exactly

gupadhyaya commented on execution.go:
```diff
@@ -73,232 +73,165 @@ func NewEngineAPIExecutionClient(
 		&engine.PayloadAttributes{
 			Timestamp:             uint64(timestamp.Unix()), //nolint:gosec // disable G115
-			Random:                c.derivePrevRandao(height),
+			Random:                c.derivePrevRandao(blockHeight),
```
if you decide to use blockHash in the InitChain, this also could use the block hash. Otherwise, `c.derivePrevRandao(0)` can be used in the InitChain. Either way is good.

tzdybal replied:
`derivePrevRandao` is completely deterministic, it's not random at all. There's an follow-up issue for this: https://github.com/rollkit/go-execution-evm/issues/34

tzdybal committed:
feat: improve initial block handling, for rollkit compatibility

tzdybal committed:
fix: handle rollkit restarts

When GetTxs is called, but payloadID is unknown, new payload is generated. This happens if GetTxs was called without previous calls to InitChain/ExecuteTxs - this happens when rollkit is restarted.

tzdybal committed:
test: add manual transaction submission test for Ethereum client

This test allows for manual submission of transactions to an Ethereum client, primarily for testing purposes. It includes nonce calculation and transaction broadcasting to assist with debugging and validation workflows. The test is skipped by default to avoid interference with automated test runs.

tzdybal committed:
chore: fix CI

tzdybal commented on execution.go:
```diff
@@ -73,232 +73,216 @@ func NewEngineAPIExecutionClient(
-	err := c.ethClient.Client().CallContext(ctx, &result, "txpool_content")
+	var payloadResult engine.ExecutionPayloadEnvelope
+	err := c.engineClient.CallContext(ctx, &payloadResult, "engine_getPayloadV3", c.payloadID)
+	c.payloadID = nil
```
Even if the call failed, we can't be sure that payloadID is still valid. It's safer to start new payload building process and get fresh payload. Otherwise code might stuck with outdated payloadID.

tzdybal committed:
chore: Remove unused mock implementations in tests

The mock implementations in `mocks_test.go` were no longer in use and have been fully removed. This cleanup simplifies the codebase and eliminates unneeded maintenance of outdated test structures.

tzdybal committed:
test: Fix loop condition in execution test to prevent overshooting

Corrects the loop condition to ensure the iteration ends at blockHeight 10 instead of exceeding it. This resolves potential out-of-bounds issues when accessing payloads during testing.

tzdybal commented on execution_test.go:
```diff
@@ -2,215 +1,428 @@
+			for i := range txs {
+				submitTransaction(t, txs[i])
+			}
+			time.Sleep(1000 * time.Millisecond)
+
```
It's just a test. 1 second simulates short block times. Not going to change this. KISS.
